### ISSUE TITLE
Overwriting of files on arm systems

### ISSUE THREAD

**Issue by mynci** [ATLauncher/ATLauncher#397]
The following describes an issue in the logic for determining whether a file should be downloaded for a particular architecture. In my case an x86 file is being downloaded for an arm system. I am using a Raspberry Pi 4 in order to run ATLauncher, it is a 64bit OS but is running a 32bit version of Java.

I am aware of the issue:
https://github.com/ATLauncher/ATLauncher/issues/390#issuecomment-658688398

To side step it I have built getMemory-linux using the Raspberry pi. This runs (albeit does not seem to return any useful data).

I am finding that no matter what I do ATLauncher overwrites my version of the file on start, command line options are ineffectual at stopping the download.

Therefore I started digging a bit deeper and I believe there may be some logic errors in how ATLaunchers decides whether or not to download a file. I am not a Java programmer (C++/Python mainly), so I apologise in advance if this is due to me misreading the code.

Looking at the logged output here:
https://paste.atlauncher.com/view/5a05c614

I could see that we were downloading a files.json:
https://download.nodecdn.net/containers/atl/launcher/json/files.json

This has the entry:
`
{"name":"getMemory-linux","folder":"tools","size":1773330,"sha1":"e5d2f326a9571d43939ae86de9c115b6c75e7da8","arch":"x86","os":"linux"}
`
Which seems to be the information used in triggering the download. This seems odd that this would even be downloaded as my arch should be arm32/aarch64/arm64. So assuming the check is made on the OS type and arch I should not get a copy of this file (I assume). However, as the log shows I do get a copy.

I believe that the following is used to determine whether a file is meant for the particular architecture:

```java
public boolean isForArchAndOs() {
        if (os != null) {
            if (os.equalsIgnoreCase("osx") && !OS.isMac()) {
                return false;
            }

            if (os.equalsIgnoreCase("windows") && !OS.isWindows()) {
                return false;
            }

            if (os.equalsIgnoreCase("linux") && !OS.isLinux()) {
                return false;
            }
        }

        if (arch != null) {
            return (arch.equalsIgnoreCase("arm64") && (!OS.is64Bit() || !OS.isArm()))
                    || (arch.equalsIgnoreCase("arm") && (OS.is64Bit() || !OS.isArm()))
                    || (arch.equalsIgnoreCase("x86") && !OS.is64Bit())
                    || (arch.equalsIgnoreCase("x64") && OS.is64Bit());
        }

        return true;
}
```
This makes perfect sense for the case where the file is for a particular OS and there is no arch specified. But it seems the logic is not quite right for my case:
The file identifies as "x86", therefore the following clause is relevant:
```java
(arch.equalsIgnoreCase("x86") && !OS.is64Bit())
```
Clearly this only looks at whether the OS is 64bit and not whether it is ARM or not.

For this clause, I believe it should read:
```java
(arch.equalsIgnoreCase("x86") && !(OS.is64Bit()||OS.isArm()))
```
I would take a guess that the following might be close to correct:
```java
        if (arch != null) {
            return (arch.equalsIgnoreCase("arm64") && (OS.is64Bit() && OS.isArm()))
                    || (arch.equalsIgnoreCase("arm") && (!OS.is64Bit() && OS.isArm()))
                    || (arch.equalsIgnoreCase("x86")  !(OS.is64Bit()||OS.isArm()))
                    || (arch.equalsIgnoreCase("x64")  !(!OS.is64Bit()||OS.isArm()))
        }
```

As I say, I thought it might be worth pointing out the issue, apologies if I have misunderstood.

If there is any quick workaround to help me to stop the code overwriting my version of the file, then please let me know. I am avoiding building from source, but will do if that's the best way to get this working.

Many thanks.

**Comment by RyanTheAllmighty** [comment]
Yes you're completely right.

Thanks for the report. I need to get my pi zero out and play around to completely confirm everything, just a hassle as I need to find my otg cable which I have no idea where it is :(

The launcher will always overwrite the "system" files from files.json regardless of any command line argument.

**Comment by RyanTheAllmighty** [comment]
Released with version 3.4.1.0
